common/membership/ringpop/factory.go

```diff
@@ -251,19 +251,27 @@ func (factory *factory) getHostInfoProvider() (membership.HostInfoProvider, erro
 		return nil, err
 	}
 
-	servicePort, ok := factory.ServicePortMap[factory.ServiceName]
+	serviceHost, ok := factory.ServicePortMap[factory.ServiceName]
 	if !ok {
 		return nil, membership.ErrUnknownService
 	}
 
 	// The broadcastAddressResolver returns the host:port used to listen for
 	// ringpop messages. We use a different port for the service, so we
 	// replace that portion.
-	serviceAddress, err := replaceServicePort(address, servicePort)
+	serviceAddress, err := extractHostFromAddress(address)
 	if err != nil {
 		return nil, err
 	}
 
-	hostInfo := membership.NewHostInfoFromAddress(serviceAddress)
+	hostInfo := membership.NewHostInfoFromServiceHost(serviceAddress, serviceHost)
 	return membership.NewHostInfoProvider(hostInfo), nil
 }
+
+func extractHostFromAddress(address string) (string, error) {
```

Comment:
we could call this function from the `replaceServicePort` function in common/membership/ringpop/monitor.go - they both return the same error if `SplitHostPort` fails. alternately, `replaceSerivcePort` is only used in exactly one other case - maybe just use this new function and a `net.JoinHostPort`?